firedrake/supermeshimpl.pyx

```diff
@@ -98,3 +103,50 @@ def assemble_mixed_mass_matrix(V_A, V_B, candidates,
 
     CHKERR(MatAssemblyBegin(mat.mat, MAT_FINAL_ASSEMBLY))
     CHKERR(MatAssemblyEnd(mat.mat, MAT_FINAL_ASSEMBLY))
+
+
+def intersection_finder(mesh_A, mesh_B):
+    # Plan:
+    # Call libsupermesh_sort_intersection_finder_set_input
+    # Call libsupermesh_sort_intersection_finder_query_output
+    # Call libsupermesh_sort_intersection_finder_get_output
+    # Return the output
+
+    cdef:
+        numpy.ndarray[long, ndim=2, mode="c"] vertex_map_A, vertex_map_B
+        numpy.ndarray[double, ndim=2, mode="c"] vertices_A, vertices_B
+        long nindices
+        numpy.ndarray[long, ndim=1, mode="c"] indices, indptr
+        long nnodes_A, nnodes_B, ncells_A, ncells_B
+        int dim_A, dim_B, loc_A, loc_B
+
+    vertices_A = mesh_A.coordinates.dat.data_ro_with_halos
+    vertices_B = mesh_B.coordinates.dat.data_ro_with_halos
+    vertex_map_A = mesh_A.coordinates.cell_node_map().values_with_halo.astype(long)
+    vertex_map_B = mesh_B.coordinates.cell_node_map().values_with_halo.astype(long)
+    nnodes_A = mesh_A.num_vertices()
+    nnodes_B = mesh_B.num_vertices()
+    dim_A = mesh_A.geometric_dimension()
+    dim_B = mesh_B.geometric_dimension()
+    ncells_A = mesh_A.num_cells()
+    ncells_B = mesh_B.num_cells()
+    loc_A = vertex_map_A.shape[1]
+    loc_B = vertex_map_B.shape[1]
+
+    libsupermesh_tree_intersection_finder_set_input(&nnodes_A, &dim_A, &ncells_A, &loc_A,
+                                                    &nnodes_B, &dim_B, &ncells_B, &loc_B,
+                                                    <const double*>vertices_A.data, <const long*>vertex_map_A.data, <const double*>vertices_B.data, <const long*>vertex_map_B.data)
```

Comment:
Needs to check that mesh_A.comm and mesh_B.comm are congruent, and pass things into libsupermesh appropriately. If libsupermesh only works over COMM_WORLD then options are (in increasing order of preference)

1. Check that we're running on COMM_WORLD, raise NotImplementedError if not
2. Fix libsupermesh and pass the communicator in.

Reply:
This intersection finding is done purely locally (no parallelism or communication), so libsupermesh doesn't see or need a communicator.

What do you mean by 'congruent' -- that they are the same (up to duplication)? I don't think this is necessary for the correct operation of the algorithm. For example, suppose that the target mesh was distributed across N cores, while the source mesh was associated with a coarse problem computed on COMM_SELF on each process independently. You can sensibly do a Galerkin projection from source to target, and need to calculate the intersections as done here, but the communicators aren't congruent (in the sense above).

What's actually required is that each core sees all the data it needs in the source mesh to compute its piece of the target function.

Reply:
OK, thanks.